### PR TITLE
CSS: Support class/elements names with colons and fix HTML character entity completion

### PR DESCRIPTION
**Pull request by matthiasblaesing**
Demo of classes with colons:

image

You can notice, that there is no error notice on the class list, the code completion shows the unescaped variant of the class and the insert (first value was inserted using completion) is done correctly.

The partitially broken behavior of character entity completion (Completion of `&am` is completed to `&a&amp;`):

image

was fixed to work correctly:

image

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
@junichi11 @KacerCZ  could you please have a look at this PR? I have a serious problem with one of the `php.editor` unittests.

For a quick glance, you can have a look here: https://github.com/apache/netbeans/actions/runs/4386975881

There are two problems I don't understand:

1. The reformatted contents differs between the two platforms (compare actual reformatted result for Linux and Windows)
2. Both unittest results do not match the result I get when running the build interactively:

image

The problem in the unittest can be reproduced locally by running:

```
ant -f php/php.editor -Dtest.type=unit -Dtest.class=org.netbeans.modules.php.editor.js.JsFormatterEmbeddedTest -Dtest.methods=testEmbeddedMultipleSections1  test-method
```

Running the test multiple times gives me sometimes the windows and sometimes the linux result. Something is fishy here, especially that this looks like a heisenbug.

Changes in the indention are kind of expected. The formatting in the test relied on the CSS indenter having run. This was the case because the `id="source"` attribute in the `script` attribute was recognized as CSS. This behavior was modified, the id attribute is parsed and handled as a normal html attribute. You can observe differing idention when you test master and replace `id=` with `ic=`. The latter will not trigger the special casing.

Any ideas are welcome.

**Comment by junichi11** [comment]
@matthiasblaesing Is it resolved?

**Comment by matthiasblaesing** [comment]
@junichi11 yes, it turns out, that this is indeed a race condition. I updated the golden file again and now in about 50% of all cases the test succeeds. The github actions run failed the first time around and succeeded on the second round.

<details>
<summary>I sprinkled some `System.out.println` into the affected indenters and then compared the output (Expand this section to see them)</summary>

```diff
diff --git a/ide/html.editor/src/org/netbeans/modules/html/editor/indent/HtmlIndentTask.java b/ide/html.editor/src/org/netbeans/modules/html/editor/indent/HtmlIndentTask.java
index 2b2427176d99..0f482dad7fb6 100644
--- a/ide/html.editor/src/org/netbeans/modules/html/editor/indent/HtmlIndentTask.java
+++ b/ide/html.editor/src/org/netbeans/modules/html/editor/indent/HtmlIndentTask.java
@@ -42,6 +42,7 @@ public class HtmlIndentTask implements IndentTask, Lookup.Provider {
     }
 
     public void reindent() throws BadLocationException {
+        System.out.println("HtmlIndentTask#reindent");
 //        long st = System.currentTimeMillis();
 //        getFormatter().process(context);
 //        Logger.getLogger("TIMER").log(Level.FINE, "HTML Reindent",
diff --git a/ide/web.indent/src/org/netbeans/modules/web/indent/api/support/AbstractIndenter.java b/ide/web.indent/src/org/netbeans/modules/web/indent/api/support/AbstractIndenter.java
index 9a41f0c28014..ddc4a0feb8ad 100644
--- a/ide/web.indent/src/org/netbeans/modules/web/indent/api/support/AbstractIndenter.java
+++ b/ide/web.indent/src/org/netbeans/modules/web/indent/api/support/AbstractIndenter.java
@@ -187,6 +187,8 @@ public abstract class AbstractIndenter<T1 extends TokenId> {
 
     private boolean used = false;
     public final void reindent() {
+        System.out.println(getClass().getSimpleName() + "#reindent");
+
         if (used && DEBUG) {
             System.err.println("WARNING: indentation task cannot be reused! is this ok?");
             //IllegalStateException x = new IllegalStateException("indentation task cannot be reused");
diff --git a/php/php.editor/src/org/netbeans/modules/php/editor/indent/PHPFormatter.java b/php/php.editor/src/org/netbeans/modules/php/editor/indent/PHPFormatter.java
index 82c9712875f6..a8d8b13ec10a 100644
--- a/php/php.editor/src/org/netbeans/modules/php/editor/indent/PHPFormatter.java
+++ b/php/php.editor/src/org/netbeans/modules/php/editor/indent/PHPFormatter.java
@@ -52,6 +52,7 @@ public class PHPFormatter implements Formatter {
 
     @Override
     public void reindent(final Context context) {
+        System.out.println("PHPFormatter#reindent");
         String mimeType = getMimeTypeAtOffset(context.document(), context.startOffset());
         String mimePath = context.mimePath(); // avoid to call twice
         if (FileUtils.PHP_MIME_TYPE.equals(mimeType) && FileUtils.PHP_MIME_TYPE.equals(mimePath)) {
@@ -62,6 +63,7 @@ public class PHPFormatter implements Formatter {
 
     @Override
     public void reformat(Context context, ParserResult info) {
+        System.out.println("PHPFormatter#reformat");
         long start = System.currentTimeMillis();
 
         (new TokenFormatter()).reformat(context, info);
diff --git a/webcommon/javascript2.editor/src/org/netbeans/modules/javascript2/editor/formatter/JsFormatter.java b/webcommon/javascript2.editor/src/org/netbeans/modules/javascript2/editor/formatter/JsFormatter.java
index 4cc67d995447..0464a7932be2 100644
--- a/webcommon/javascript2.editor/src/org/netbeans/modules/javascript2/editor/formatter/JsFormatter.java
+++ b/webcommon/javascript2.editor/src/org/netbeans/modules/javascript2/editor/formatter/JsFormatter.java
@@ -101,6 +101,7 @@ public class JsFormatter implements Formatter {
 
     @Override
     public void reformat(final Context context, final ParserResult compilationInfo) {
+        System.out.println("JsFormatter#reformat");
         processed.clear();
         lastOffsetDiff = 0;
         final Document doc = context.document();
@@ -1909,6 +1910,8 @@ public class JsFormatter implements Formatter {
     @Override
     public void reindent(final Context context) {                                                                                                                                      
                                                                                                                                                                                        
+        System.out.println("JsFormatter#reindent");                                                                                                                                    
+                                                                                                                                                                                       
         Document document = context.document();                                                                                                                                        
         int startOffset = context.startOffset();                                                                                                                                       
         int endOffset = context.endOffset();
```
</details>

With that I get:

```
Fail: 
    [junit] HtmlIndentTask#reindent
    [junit] HtmlIndenter#reindent
    [junit] CssIndenter#reindent
    [junit] JsFormatter#reformat
    [junit] PHPFormatter#reformat

Ok:
    [junit] HtmlIndentTask#reindent
    [junit] HtmlIndenter#reindent
    [junit] JsFormatter#reformat
    [junit] CssIndenter#reindent
    [junit] PHPFormatter#reformat
```

So indeed the order of `CssIndenter` and `JsFormatter` is different. Next step is to see how the order of identers is determined.

**Comment by matthiasblaesing** [comment]
I updated this PR with a fix for the changing order of the language paths. A more complete explanation can be found in the commit message itself:

https://github.com/apache/netbeans/pull/5641/commits/be093c3486ff0b9858aea8b2274ca646f87d48c6

It is hacky, but at least more predictable than in the past.

**Comment by junichi11** [comment]
@matthiasblaesing Makes sense. Great!

File: ide/css.editor/src/org/netbeans/modules/css/editor/module/spi/CssCompletionItem.java [hunk 1]
```diff
@@ -609,6 +620,42 @@ public ElementKind getKind() {
...
             return super.getSortPrioOverride() - (related ? 1 : 0);
         }
+
+        /**
+         * Escape the input for usage in CSS files.
+         *
+         * @param input
+         * @return
+         */
+        private static String escape(String input) {
```
Comment (mbien) [inline ide/css.editor/src/org/netbeans/modules/css/editor/module/spi/CssCompletionItem.java:630]: shouldn't NB use some encoding library, e.g. "OWASP Java Encoder" is fairly popular (and tiny).
https://github.com/OWASP/owasp-java-encoder
https://javadoc.io/static/org.owasp.encoder/encoder/1.2.3/org/owasp/encoder/Encode.html

Comment (matthiasblaesing) [reply ide/css.editor/src/org/netbeans/modules/css/editor/module/spi/CssCompletionItem.java:630]: I would agree if we would talk about a web application, but in this case the usage is not security, but compatibility. 

I saw one thing in the description of the CSS escaper of owasp, that looks like a potential problem for my approach. I overlooked that the hexadecimal encoding and decoding needs a bounding whitespace or can have it. I'll revisit this and the CSS grammar.

**Comment by matthiasblaesing** [comment]
@mbien I pushed an update with improved CSS escaping support and added unittests to verify the behavior. The intention is to squash that final commit into the origin "colon support" commit.

**Comment by mbien** [comment]
@matthiasblaesing oh yeah I totally overlooked the decoding part here :). Its just that I did see a bug in code like that before and the solution there was to simply switch to a JDK provided class (was for urls or something similar, I don't think I will be able to find the PR now).

commons text unfortunately doesn't handle CSS:
https://commons.apache.org/proper/commons-text/javadocs/api-release/org/apache/commons/text/StringEscapeUtils.html

I can't come up with any other lib right now, so it is probably fine to use custom code.

**Comment by mbien** [comment]
found this: https://github.com/unbescape/unbescape (never used it before)

**Comment by matthiasblaesing** [comment]
Thank you for the link to `unbescape` it is an interesting library. At this point in time I'll stick with the handcrafted version. That way I know what can be done to customize this and what not. External libraries also cause overhead (shipping, licensing, updating, incompatible changes, missing functions) and currently I think it is not worth that.

**Review by NicolaIsotta — APPROVED** [review]

File: ide/html.editor/src/org/netbeans/modules/html/editor/CssPreferences.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.prefs.Preferences;
+import org.netbeans.api.editor.mimelookup.MimeLookup;
+import org.netbeans.modules.css.lib.api.CssTokenId;
+import org.openide.util.WeakListeners;
+
+/**
+ *
+ * @author mfukala@netbeans.org
+ */
+public class CssPreferences {
```
Comment (mbien) [inline ide/html.editor/src/org/netbeans/modules/html/editor/CssPreferences.java:33]: where did this class come from? I suppose a copy from another module, just renamed?

Comment (matthiasblaesing) [reply ide/html.editor/src/org/netbeans/modules/html/editor/CssPreferences.java:33]: It is a reduced copy from `css.editor`. I neither want to add more to the exposed API nor make `html.editor` using an implementation dependency.

Comment (mbien) [reply ide/html.editor/src/org/netbeans/modules/html/editor/CssPreferences.java:33]: understood

**Pull request merged**